crates/turbopack-dev-server/src/source/route_tree.rs

```diff
@@ -0,0 +1,372 @@
+use std::{fmt::Write, mem::replace};
+
+use anyhow::Result;
+use indexmap::IndexMap;
+use serde::{Deserialize, Serialize};
+use turbo_tasks::{
+    primitives::StringVc, trace::TraceRawVcs, TaskInput, TryJoinIterExt, ValueToString,
+    ValueToStringVc,
+};
+
+use super::{GetContentSourceContentVc, GetContentSourceContentsVc};
+
+#[derive(TaskInput, Clone, Debug, PartialEq, Eq, Serialize, Deserialize, TraceRawVcs)]
+pub enum FinalSegment {
+    CatchAll,
+    Fallback,
+    NotFound,
+}
+
+#[derive(TaskInput, Clone, Debug, PartialEq, Eq, Serialize, Deserialize, TraceRawVcs)]
+pub enum BaseSegment {
+    Static(String),
+    Dynamic,
+}
+
+impl BaseSegment {
+    pub fn from_static_pathname(str: &str) -> impl Iterator<Item = BaseSegment> + '_ {
+        str.split('/')
+            .filter(|s| !s.is_empty())
+            .map(|s| BaseSegment::Static(s.to_string()))
+    }
+}
+
+#[turbo_tasks::value(transparent)]
+pub struct RouteTrees(Vec<RouteTreeVc>);
+
+#[turbo_tasks::value_impl]
+impl RouteTreesVc {
+    #[turbo_tasks::function]
+    pub async fn merge(self) -> Result<RouteTreeVc> {
+        let trees = &*self.await?;
+        if trees.is_empty() {
+            return Ok(RouteTree::default().cell());
+        }
+        if trees.len() == 1 {
+            return Ok(*trees.iter().next().unwrap());
+        }
+
+        // Find common base
+        let mut tree_values = trees.iter().try_join().await?;
+        let mut common_base = 0;
+        let last_tree = tree_values.pop().unwrap();
+        'outer: while common_base < last_tree.base.len() {
+            for tree in tree_values.iter() {
+                if tree.base.len() <= common_base {
+                    break 'outer;
+                }
+                if tree.base[common_base] != last_tree.base[common_base] {
+                    break 'outer;
+                }
+            }
+            common_base += 1;
+        }
+        tree_values.push(last_tree);
```

Comment:
Did we need to pop it in the first place?

Reply:
Technically not, but without that we would need to skip the last element of an iterator, which has more code.